### PR TITLE
Flax NNX GSPMD guide

### PR DESCRIPTION
**Pull request by IvyZX**
Add a guide to do GSPMD-style sharding annotation on NNX models.

Covered everything in the [Linen pjit guide](https://flax.readthedocs.io/en/latest/guides/parallel_training/flax_on_pjit.html), but better explanations and demonstrations, and more concise code!

Also added a small example for loading sharded model from checkpoint.

[Preview](https://flax-nnx--4220.org.readthedocs.build/en/4220/guides/flax_gspmd.html)

### CODE REVIEW

File: docs_nnx/guides/flax_gspmd.md [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        rngs.params(), (depth, depth))   # RNG key and shape for W2 creation
+    )
+
+  def __call__(self, x: jax.Array):
+    y = self.dot1(x)
+    y = jax.nn.relu(y)
+    # In data parallelism, input / intermediate value's first dimension (batch)
+    # will be sharded on `data` axis
+    y = jax.lax.with_sharding_constraint(y, PartitionSpec('data', 'model'))
+    z = jnp.dot(y, self.w2.value)
```
Comment (cgarciae) [inline docs_nnx/guides/flax_gspmd.md:105]: Variables can be used as JAX arrays thanks to the  `__jax_aray__` protocol.
```suggestion
    z = jnp.dot(y, self.w2)
```

File: docs_nnx/guides/flax_gspmd.md [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+# We have annotations sticked there, yay!
+print(unsharded_model.dot1.kernel.sharding)     # (None, 'model')
+print(unsharded_model.w2.sharding)              # ('model', None)
+
+# But the actual arrays are not sharded... wut?
+print(unsharded_model.dot1.kernel.value.sharding)  # SingleDeviceSharding
+print(unsharded_model.w2.value.sharding)           # SingleDeviceSharding
+```
+
+We should leverage JAX's compilation mechanism, aka. `jax.jit`, to create the sharded model. The key is to intialize a model and assign shardings upon the model state within a jitted function:
```
Comment (cgarciae) [inline docs_nnx/guides/flax_gspmd.md:138]:
```suggestion
We should leverage JAX's compilation mechanism, via `nnx.jit`, to create the sharded model. The key is to intialize a model and assign shardings upon the model state within a jitted function:
```

File: docs_nnx/guides/flax_gspmd.md [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+1. Use [`nnx.get_partition_spec`](https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/spmd.html#flax.nnx.get_partition_spec) to strip out the `.sharding` annotations attached upon model variables;
+
+1. Call [`jax.lax.with_sharding_constraint`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.with_sharding_constraint.html) to bind the model state with the sharding annotations. This API tells the top-level `jax.jit` how to shard a variable!
+
+1. Throw away the unsharded state and return the model based upon the sharded state.
+
+1. Compile the whole function with `nnx.jit` instead of `jax.jit` because it allows the output to be a stateful NNX module.
+
+1. Run it under a device mesh context so that JAX knows which devices to shard it to.
```
Comment (cgarciae) [inline docs_nnx/guides/flax_gspmd.md:148]:
```suggestion
1. Call [`jax.lax.with_sharding_constraint`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.with_sharding_constraint.html) to bind the model state with the sharding annotations. This API tells the top-level `nnx.jit` how to shard a variable!

1. Throw away the unsharded state and return the model based upon the sharded state.

1. Run it under a device mesh context so that JAX knows which devices to shard it to.
```

File: docs_nnx/guides/flax_gspmd.md [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+                                      args=ocp.args.StandardRestore(abs_state))
+jax.debug.visualize_array_sharding(loaded_sharded.dot1.kernel.value)
+jax.debug.visualize_array_sharding(loaded_sharded.w2.value)
+```
+
+## Compile the training loop
+
+Now, from initialization or from checkpoint, we have a sharded model. To carry out the compiled, scaled up training, we need to shard the inputs as well. In this data parallelism example, the training data has its batch dimension sharded across `data` device axis, so you should put your data in sharding `('data', None)`. You can use `jax.device_put` for this.
+
+Note that with the correct sharding for all inputs, the output will be sharded in the most natural way even without `jax.jit`. See the example below - even without `jax.lax.with_sharding_constraint` on the output `y`, it was still sharded as `('data', None)`.
```
Comment (cgarciae) [inline docs_nnx/guides/flax_gspmd.md:233]:
```suggestion
Note that with the correct sharding for all inputs, the output will be sharded in the most natural way even without `nnx.jit`. See the example below - even without `jax.lax.with_sharding_constraint` on the output `y`, it was still sharded as `('data', None)`.
```

File: docs_nnx/guides/flax_gspmd.md [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+def block_all(xs):
+  jax.tree_util.tree_map(lambda x: x.block_until_ready(), xs)
+  return xs
+
+with mesh:
+  new_state = block_all(train_step(sharded_model, optimizer, input, label))
+```
+
+## Logical axis annotation
```
Comment (cgarciae) [inline docs_nnx/guides/flax_gspmd.md:290]: Nice! I didn't know about `sharding_rules`. In `nnx_lm1b` with have this other pattern which maps the mesh axes in the constructor:
- https://github.com/google/flax/blob/e3772b2d01ba8aca7a09644ed66d4bc072c8782b/examples/lm1b_nnx/configs/default.py#L22
- https://github.com/google/flax/blob/e3772b2d01ba8aca7a09644ed66d4bc072c8782b/examples/lm1b_nnx/models.py#L214

**Review by cgarciae — APPROVED** [review]

File: docs_nnx/guides/flax_gspmd.md [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+    jupytext_version: 1.13.8
+---
+
+# Scale up on multiple devices
+
+This guide shows how to scale up [Flax NNX Modules](https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/module.html#flax.nnx.Module) on multiple devices and hosts using [`jax.jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html).
+
++++
+
+## Flax and `jax.jit` scaled up
```
[the remark this answers is not in the document]
Comment (IvyZX) [reply docs_nnx/guides/flax_gspmd.md:17]: I kinda want to convey the idea that essentially we are using JAX's compilation machinery for the scaling up work. I renamed the title and added another paragraph explaining this (and mentioning `nnx.jit` there).

File: docs_nnx/guides/flax_gspmd.md [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+      depth, depth,
+      kernel_init=nnx.with_partitioning(init_fn, (None, 'model')),
+      use_bias=False,  # or use `bias_init` to give it annotation too
+      rngs=rngs)
+
+    # Initialize a weight param `w2` and annotate with sharding ('model', None)
+    self.w2 = nnx.Param(
+      nnx.with_partitioning(init_fn, ('model', None))(
+        rngs.params(), (depth, depth))   # RNG key and shape for W2 creation
+    )
```
[the remark this answers is not in the document]
Comment (IvyZX) [reply docs_nnx/guides/flax_gspmd.md:97]: Good idea!

File: docs_nnx/guides/flax_gspmd.md (hunk 1 above)
Comment (IvyZX) [reply docs_nnx/guides/flax_gspmd.md:105]: Done.

File: docs_nnx/guides/flax_gspmd.md (hunk 2 above)
Comment (IvyZX) [reply docs_nnx/guides/flax_gspmd.md:138]: Done.

File: docs_nnx/guides/flax_gspmd.md (hunk 3 above)
Comment (IvyZX) [reply docs_nnx/guides/flax_gspmd.md:148]: Hmm... I think we should still briefly explain why using `nnx.jit` is a better pattern. Especially since we are making transforms closer to JAX style now we should assume some users have experience with `jax.jit`. I can remove the mentions of `jax.jit` here and direct users more explicitly to `nnx.jit`.

File: docs_nnx/guides/flax_gspmd.md (hunk 4 above)
Comment (IvyZX) [reply docs_nnx/guides/flax_gspmd.md:233]: Done.

File: docs_nnx/guides/flax_gspmd.md (hunk 5 above)
Comment (IvyZX) [reply docs_nnx/guides/flax_gspmd.md:290]: Yeah I added them recently to align with Linen's `LogicallyPartitioned`. It's just annotations so there's actually a ton of ways to make them work, and I like how you made it in `nnx_lm1b`!
